iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,172 @@
-        HTML.  Tags that are required to be empty (e.g. <br>) are left as
-        self-closing.
-        """
-        for e in xml.iter('*'):
-            m = re.match(r'\{http://www\.w3\.org/1999/xhtml\}(.*)', e.tag)
-            if (m is None or m.group(1) not in XHTMLSerializer.selfClosableElements) and e.text is None:
-                e.text = ''
-
-    def serialize(self, xmlDocument, fout):
-        self._expandEmptyTags(xmlDocument)
-        fout.write(etree.tostring(xmlDocument, method="xml", encoding="utf-8", xml_declaration=True))
+    ENTITIES = {
+        '>': '&gt;',
+        '<': '&lt;',
+        '"': '&quot;',
+        '&': '&amp;'
+        }
+
+    def __init__(self, fout, xml_declaration = True, assume_xhtml = True):
+        self.fout = fout
+        self.encoding = "utf-8"
+        self.xml_declaration = xml_declaration
+        self.assume_xhtml = assume_xhtml
+
+    def write(self, s):
+        self.fout.write(s.encode(self.encoding))
+
+    def prefix_sort(self, p):
+        return p if p is not None else '0'
+
+    def qname_for_node(self, node):
+        qname = etree.QName(node.tag)
+        if qname.namespace is None:
+            prefix = None
+        elif node.nsmap.get(node.prefix, None) == qname.namespace:
+            prefix = node.prefix
+        else:
+            prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace), key = self.prefix_sort)))
+        if prefix is None:
+            return qname.localname
+        return "%s:%s" % (prefix, qname.localname)
+
+    def qname_for_attr(self, tag, nsmap):
+        qname = etree.QName(tag)
+        if qname.namespace is None:
+            return qname.localname
+        if qname.namespace == XML_NS:
+            prefix = 'xml'
+        else:
+            prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace and p is not None), key = self.prefix_sort)))
+        return "%s:%s" % (prefix, qname.localname)
+
+    def is_selfclosable(self, n):
+        qname = etree.QName(n)
+        return (
+                qname.localname in XHTMLSerializer.SELF_CLOSABLE
+                and (
+                    qname.namespace == XHTML_NS
+                    or (self.assume_xhtml and qname.namespace is None)))
+
+    def escape_attr(self, s):
+        return re.sub(r'([<>"&\u0001-\u001F\u007F-\u009F])', lambda m: self.escape_char(m.group(0)), s)
+
+    def xmlns_declaration(self, prefix, uri):
+        if prefix is None:
+            return 'xmlns="%s"' % self.escape_attr(uri)
+        return 'xmlns:%s="%s"' % (prefix, self.escape_attr(uri))
+
+    def namespace_declarations(self, new_nsmap, cur_nsmap):
+        changed = set(p[0] for p in (set(new_nsmap.items()) ^ set(cur_nsmap.items())))
+        return sorted(self.xmlns_declaration(p, new_nsmap[p]) for p in changed)
+
+    def escape_char(self, c):
+        return self.ENTITIES.get(c, '&#x%2X;' % ord(c))
```

Comment:
I think remove the padding:
```
$ echo $'<r>\n</r>' | python b.py | xmllint -
-:2: parser error : CharRef: invalid hexadecimal value
<r>&#x A;</r>
      ^
```
Well, it might be worth writing out at least \t\r\n directly.

Reply:
Oh dear, this commit was a mess!  I was aiming for `%02X` so that it's zero padded, and I misread the bit about `0x01` to `0x1F` being newly allowed in 1.1 as them _all_ being previously forbidden in 1.0, when obviously that can't be right.  And all the multi-line tests that we have so far neatly dodge this code because they're only testing normalisation of whitespace outside the root element.

Reply:
This was done in c3c45ad44a366a7150cda6dc0034a51b9915eb2a